### PR TITLE
Jenkins CI/CD piplines

### PR DESCRIPTION
**Pull request by PeterJausovec**
Creating CI/CD piplines in Jenkins using proper naming

### CODE REVIEW

File: src/command_modules/azure-cli-project/azure/cli/command_modules/project/custom.py [hunk 1]
```diff
@@ -75,7 +75,7 @@ def browse_pipeline():
             'Jenkins host name does not exist in projectSettings.json')
 
     local_port = _get_available_local_port()
-    local_address = 'http://127.0.0.1:{}'.format(local_port)
+    local_address = 'http://127.0.0.1:{}/blue'.format(local_port)
```
Comment (saurabsa) [inline src/command_modules/azure-cli-project/azure/cli/command_modules/project/custom.py]: >/blue [](start = 40, length = 5)

something hardcoded for a service or is it common to all services?

File: src/command_modules/azure-cli-project/azure/cli/command_modules/project/custom.py [hunk 2]
```diff
@@ -240,16 +242,108 @@ def create_deployment_pipeline(remote_access_token):  # pylint: disable=unused-a
...
+
+            jenkins_hostname = _get_remote_host(
+                jenkins_resource.dns_prefix, jenkins_resource.location)
+            project_settings.jenkins_hostname = jenkins_hostname
+            project_settings.set_ci_pipeline_name(
+                jenkins_resource._get_ci_job_name(), _get_git_remote_url(), _get_service_folder())
+            project_settings.set_cd_pipeline_name(
+                jenkins_resource._get_cd_job_name(), _get_git_remote_url(), _get_service_folder())
+
+            utils.writeline('Created')
```
Comment (saurabsa) [inline src/command_modules/azure-cli-project/azure/cli/command_modules/project/custom.py]: >writeline [](start = 18, length = 9)

Where do we want to show Created ? On a new line? If 'Creating build and deployment pipelines for {} .............. Created' looks good, please replace this with write()

File: src/command_modules/azure-cli-project/azure/cli/command_modules/project/custom.py [hunk 3]
```diff
@@ -240,16 +242,108 @@ def create_deployment_pipeline(remote_access_token):  # pylint: disable=unused-a
...
+            if longprocess:
+                longprocess.process_stop()
+            current_process = None
+        else:
+            # Check if the pipelines are already created
+            if _deployment_pipelines_exist():
+                sleep(5)
+                utils.writeline('Created')
```
Comment (saurabsa) [inline src/command_modules/azure-cli-project/azure/cli/command_modules/project/custom.py]: >writeline [](start = 22, length = 9)

same

File: src/command_modules/azure-cli-project/azure/cli/command_modules/project/custom.py [hunk 4]
```diff
@@ -240,16 +242,108 @@ def create_deployment_pipeline(remote_access_token):  # pylint: disable=unused-a
...
+
+                longprocess = utils.Process()
+                current_process = longprocess
+
+                jenkins_resource.configure()
+                project_settings.set_ci_pipeline_name(
+                    jenkins_resource._get_ci_job_name(), _get_git_remote_url(), _get_service_folder())
+                project_settings.set_cd_pipeline_name(
+                    jenkins_resource._get_cd_job_name(), _get_git_remote_url(), _get_service_folder())
+                utils.writeline('Created')
```
Comment (saurabsa) [inline src/command_modules/azure-cli-project/azure/cli/command_modules/project/custom.py]: >writeline [](start = 22, length = 9)

same

File: src/command_modules/azure-cli-project/azure/cli/command_modules/project/jenkins.py [hunk 5]
```diff
@@ -5,12 +5,17 @@
...
+import azure.cli.command_modules.project.settings as settings
 from azure.cli.command_modules.project.deployments import DeployableResource
 import azure.cli.core.azlogging as azlogging  # pylint: disable=invalid-name
-logger = azlogging.get_az_logger(__name__) # pylint: disable=invalid-name
+import tempfile
+import hashlib
+logger = azlogging.get_az_logger(__name__)  # pylint: disable=invalid-name
+
+# pylint: disable=line-too-long, too-many-arguments,
+# too-many-instance-attributes
```
Comment (saurabsa) [inline src/command_modules/azure-cli-project/azure/cli/command_modules/project/jenkins.py]: I have seen this, it doesnot work. This is not just a comment. pylint uses this. All # pylint: disable= has to be in the same line. If you are breaking the line then add prefix "# pylint: disable=" in the next line as well.

File: src/command_modules/azure-cli-project/azure/cli/command_modules/project/jenkins.py [hunk 6]
```diff
@@ -51,69 +61,190 @@ def _deployment_completed(self, completed_deployment):
...
         self._unsecure_instance()
-        utils.writeline('Jenkins configuration completed.')
+        logger.info('Jenkins configuration completed.')
+
+    def _get_hostname(self):
+        """
+        Gets the Jenkins hostname
+        """
+        return '{}.{}.cloudapp.azure.com'.format(self.dns_prefix, self.location)
+
```
Comment (saurabsa) [inline src/command_modules/azure-cli-project/azure/cli/command_modules/project/jenkins.py]: I have put a similar method in utils(). You can use it after I merge or vice versa.

File: src/command_modules/azure-cli-project/azure/cli/command_modules/project/jenkins.py [hunk 7]
```diff
@@ -51,69 +61,190 @@ def _deployment_completed(self, completed_deployment):
...
+        ssh.load_system_host_keys()
+        ssh.set_missing_host_key_policy(paramiko.AutoAddPolicy())
+        ssh.connect('{}.{}.cloudapp.azure.com'.format(
+            dns_prefix, location), username=user)
+        scp = paramiko.SFTPClient.from_transport(ssh.get_transport())
+        scp.get('.kube/config', local_kube_config)
+        scp.close()
+        ssh.close()
+
+        return local_kube_config
```
Comment (saurabsa) [inline src/command_modules/azure-cli-project/azure/cli/command_modules/project/jenkins.py]: please use SSHConnect after I merge.

File: src/command_modules/azure-cli-project/azure/cli/command_modules/project/custom.py (hunk 1 above)
Comment (PeterJausovec) [inline src/command_modules/azure-cli-project/azure/cli/command_modules/project/custom.py]: this is not service specific, it's the Jenkins' new UI (Blueocean)

File: src/command_modules/azure-cli-project/azure/cli/command_modules/project/jenkins.py (hunk 7 above)
Comment (PeterJausovec) [inline src/command_modules/azure-cli-project/azure/cli/command_modules/project/jenkins.py]: I'll do that as a separate change

File: src/command_modules/azure-cli-project/azure/cli/command_modules/project/jenkins.py (hunk 6 above)
Comment (PeterJausovec) [inline src/command_modules/azure-cli-project/azure/cli/command_modules/project/jenkins.py]: I'll do that as a separate change

File: src/command_modules/azure-cli-project/azure/cli/command_modules/project/jenkins.py (hunk 5 above)
Comment (PeterJausovec) [inline src/command_modules/azure-cli-project/azure/cli/command_modules/project/jenkins.py]: fixed; linter was automatically breaking the line into two

**Pull request merged**